aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,230 @@
+
+Fortunately, .NET has abstracted validation into data annotations. These attributes contain validation code, thereby reducing the amount of code you must write. 
+
+Data annotations
+----------------
+Data annotations are a way to configure model validation so it's similar conceptually to validation on fields in tables. This includes constraints such as assigning data types or required fields. Other types of validation include applying patterns to data to enforce business rules, such as a credit card, phone number, or email address. Data annotations make enforcing these requirements much simpler and easier to use.
+
+Below is an annotated ``Movie`` model from an app that stores information about movies and TV shows. Most of the properties are required, the ``Id`` property is a primary key, and several string properties have length requirements. Additionally, there is a numeric range in place for the ``Price`` property from 0 to $999.99, along with a custom annotation.
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-34
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[CreditCard]``: Specifies the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[EmailAddress]``: Specifies the property with an email format.
+- ``[Phone]``: Specifies the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Specifies the property as a URL.
+
+MVC supports any attribute that derives from ``ValidationAttribute`` for validation purposes. Many useful validation attributes can be found in the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_.
+
+There may be intances where you need more features than built-in attributes provide. For those times, you can create custom validation attributes by deriving from ``ValidationAttribute`` or change your model to subclass ``IValidatableObject``.
+
+Model State
+-----------
+Model state represents validation errors that were submitted with HTML form values. You can then send back model state errors to the client. For example, if a movie is marked as a preorder, it should not already be released to the public. The view will display any errors discovered and added to ``ModelState`` as shown in the example below:
```

Comment:
Not sure what the last sentence means. There is no view in the sample code. And the only information from the `ModelState` that is returned is a broad indication (a BadRequest status code) that something went wrong when `!ModelState.IsValid`.

Suggest returning `HttpBadRequest(ModelState)`. That is only for use in API actions but includes a "safe" serialization of errors in a JSON (normally) response.